pkg/solana/txm/txm.go

```diff
@@ -175,193 +175,236 @@ func (txm *Txm) run() {
-			msg.cfg.ComputeUnitPriceMax,
-			msg.cfg.ComputeUnitPriceMin,
-			uint(count), //nolint:gosec // reasonable number of bumps should never cause overflow
-		)
-		return fees.ComputeUnitPrice(fee)
+	// Build and sign initial transaction setting compute unit price
+	initTx, err := txm.buildTx(ctx, msg, 0)
+	if err != nil {
+		return solanaGo.Transaction{}, "", solanaGo.Signature{}, err
 	}
 
-	baseTx := msg.tx
+	ctx, cancel := context.WithTimeout(ctx, msg.cfg.Timeout)
+	sig, err := txm.sendInitialTx(ctx, initTx, msg, cancel)
+	if err != nil {
+		return solanaGo.Transaction{}, "", solanaGo.Signature{}, err
+	}
 
-	// add compute unit limit instruction - static for the transaction
-	// skip if compute unit limit = 0 (otherwise would always fail)
-	if msg.cfg.ComputeUnitLimit != 0 {
-		if computeUnitLimitErr := fees.SetComputeUnitLimit(&baseTx, fees.ComputeUnitLimit(msg.cfg.ComputeUnitLimit)); computeUnitLimitErr != nil {
-			return solanaGo.Transaction{}, "", solanaGo.Signature{}, fmt.Errorf("failed to add compute unit limit instruction: %w", computeUnitLimitErr)
-		}
+	// Initialize signature list with initialTx signature. This list will be used to add new signatures and track retry attempts.
+	sigs := &signatureList{}
+	sigs.Allocate()
+	if initSetErr := sigs.Set(0, sig); initSetErr != nil {
+		return solanaGo.Transaction{}, "", solanaGo.Signature{}, fmt.Errorf("failed to save initial signature in signature list: %w", initSetErr)
 	}
 
-	buildTx := func(ctx context.Context, base solanaGo.Transaction, retryCount int) (solanaGo.Transaction, error) {
-		newTx := base // make copy
+	// pass in copy of msg (to build new tx with bumped fee) and broadcasted tx == initTx (to retry tx without bumping)
+	txm.done.Add(1)
+	go func() {
+		defer txm.done.Done()
+		txm.retryTx(ctx, msg, initTx, sigs)
+	}()
 
-		// set fee
-		// fee bumping can be enabled by moving the setting & signing logic to the broadcaster
-		if computeUnitErr := fees.SetComputeUnitPrice(&newTx, getFee(retryCount)); computeUnitErr != nil {
-			return solanaGo.Transaction{}, computeUnitErr
-		}
+	// Return signed tx, id, signature for use in simulation
+	return initTx, msg.id, sig, nil
+}
 
-		// sign tx
-		txMsg, marshalErr := newTx.Message.MarshalBinary()
-		if marshalErr != nil {
-			return solanaGo.Transaction{}, fmt.Errorf("error in soltxm.SendWithRetry.MarshalBinary: %w", marshalErr)
-		}
-		sigBytes, signErr := txm.ks.Sign(ctx, key, txMsg)
-		if signErr != nil {
-			return solanaGo.Transaction{}, fmt.Errorf("error in soltxm.SendWithRetry.Sign: %w", signErr)
+// prepareTransaction sets blockhash and lastValidBlockHeight which will be used to track expiration.
+// If required, it also performs balanceCheck and sets compute unit limit.
+func (txm *Txm) prepareTransaction(ctx context.Context, msg *pendingTx) error {
+	client, err := txm.client.Get()
+	if err != nil {
+		return fmt.Errorf("failed to get client in sendWithRetry: %w", err)
+	}
+	blockhash, err := client.LatestBlockhash(ctx)
```

Comment:
by moving things around we may be able to execute this rpc call only when explicitly needed to overwrite blockhash (in case of expiration). This may require to add lastValidBlockheight in Enqueue signature

Reply:
done